Convert EmbedPreview component to functional component

louwie17 opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?

This PR converts the `EmbedPreview` component to a functional component.

## Why?

React class components are outdated and functional components are encouraged, this is part of the https://github.com/WordPress/gutenberg/issues/22890 to convert all class components to functional components.

## How?
Converts `EmbedPreview` component to a functional component.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

1. Open a Post or Page
2. Add an Embed block
3. Grab a link from YouTube or Tumblr or any of the sites listed here: https://wordpress.org/documentation/article/embeds/#list-of-sites-you-can-embed-from and add it to the Embed component
4. Make sure it saves correctly and no new errors are shown in the console.
5. Do a smoke test to make sure it continues to work as before ( changing the width in the toolbar, or changing the url, clicking around in the video/image to make sure things are interactive.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
N/A

## Screenshots or screencast <!-- if applicable -->

N/A

louwie17 committed:
Convert EmbedPreview component to functional component

ntsekouras commented on packages/block-library/src/embed/embed-preview.js:
```diff
@@ -18,150 +18,131 @@ import {
-			// happens on mouseup when the overlay is clicked.
-			return { interactive: false };
+	useEffect( () => {
+		if ( isSelected && interactive ) {
```
We should also add the `interactive` dep.
```suggestion
		if ( ! isSelected && interactive ) {
```

kevin940726 replied:
Also consider just [calling `setState` during rendering](https://react.dev/reference/react/Component#static-getderivedstatefromprops) to replace `getDerivedStateFromProps`. Something like:

```js
if ( ! isSelected && interactive ) {
  setInteractive( false );
}
```

louwie17 replied:
Thanks for the suggestions and the missed negative condition. I updated it in this commit: https://github.com/WordPress/gutenberg/pull/51325/commits/f4349e489e7d6f7a91caf592218df87af57e38ac
I also rebased the PR.

ntsekouras reviewed (commented):
Thanks for the PR @louwie17! This looks good except for my small comment.

t-hamano commented:
Hi @louwie17,

I found this PR when trying to resolve some of the issues in #65179. If you have time to work on this issue, could you rebase this PR to resolve the conflicts?

Thank you!

louwie17 commented:
> Hi @louwie17,
> 
> I found this PR when trying to resolve some of the issues in #65179. If you have time to work on this issue, could you rebase this PR to resolve the conflicts?

@t-hamano I had forgotten about this PR, I should be able to resolve the conflicts.

louwie17 committed:
Update interactive condition and move call inline versus within useEffect

louwie17 commented:
> I found this PR when trying to resolve some of the issues in #65179. If you have time to work on this issue, could you rebase this PR to resolve the conflicts?

The PR is updated.

@ntsekouras, @kevin940726 sorry for the very long delay in addressing the PR feedback, but it has been addressed. This PR is ready for a re-review.

t-hamano reviewed (approved):
LGTM! Thanks for the update 👍

The overlay is also working correctly. That is, the first time the block is clicked, elements in the iframe document do not react the first time the block is clicked, but only when the block is selected:

https://github.com/user-attachments/assets/6e95c46a-41cd-43de-8ee9-7356fbb7350e

louwie17 commented:
Thanks for the quick review @t-hamano, I don't have permission to merge, so I will leave that up to you or any of the other reviewers :)
